### PR TITLE
fix: not disposing autoDispose family provider

### LINKED ISSUE #2480: Nested ProviderScope fix breaks family providers(?)
**Describe the bug**

Commit fea8a961c04c0a6ae840316fcfbd330220ff5ce1 seems to have broken a lot of providers in our app. 

For now I can say that it looks like rebuilds aren't being triggered for ConsumerState instances that are using ref.watch() in their build methods even though the watched provider is changed.

In this instance, the underlying provider is a StateNotifierProvider with `autoDispose` and `family`, where we are switching the `family` parameter as we switch pages in the app. The StateNotifier holds a `List<DataItem>?` that is initially `null`, but later updated with a value. Only the Widget on the initial "page" shows list items. Once we switch to a new page (using a new family argument) the new Widget only shows the provided value as being `null`, even though it does get updated once data is fetched (which should trigger a rebuild).

Sorry for the vagueness of this report, I'm still investigating and will provide more details if I figure anything out.

**To Reproduce**

<Please add a small sample to that can be executed to reproduce the problem. As a general rule, 100 lines is the maximum>

**Expected behavior**
ConsumerState should rebuild when watched provider changes.

### PR DESCRIPTION
**Pull request by jeiea**
Close #2480.

### CODE REVIEW

File: packages/riverpod/test/framework/auto_dispose_test.dart [hunk 1]
```diff
@@ -529,6 +529,33 @@ final alwaysAlive = Provider((ref) {
     expect(container.getAllProviderElements(), isEmpty);
   });
 
+  test(
+      'when cancel autoDispose family override it should be disposed immediately',
+      () async {
+    final provider = Provider.autoDispose.family<int, int>((ref, _) => -1);
```
Comment (rrousselGit) [inline packages/riverpod/test/framework/auto_dispose_test.dart:535]:
```suggestion
    // Regression test for https://github.com/rrousselGit/riverpod/issues/2480
    final provider = Provider.autoDispose.family<int, int>((ref, _) => -1);
```

File: packages/riverpod/test/framework/auto_dispose_test.dart [hunk 2]
```diff
@@ -529,6 +529,33 @@ final alwaysAlive = Provider((ref) {
...
+
+    var constructionCount = 0;
+    final override = Provider.autoDispose.family<int, int>((ref, _) {
+      return ++constructionCount;
+    });
+
+    final overriden = provider.overrideWithProvider(override);
```
Comment (rrousselGit) [inline packages/riverpod/test/framework/auto_dispose_test.dart:542]: Do not create variables for these :)

Comment (rrousselGit) [reply packages/riverpod/test/framework/auto_dispose_test.dart:542]: In fact you could switch to overrideWith too

File: packages/riverpod/test/framework/auto_dispose_test.dart [hunk 3]
```diff
@@ -529,6 +529,33 @@ final alwaysAlive = Provider((ref) {
...
+    final root = createContainer();
+    final container = createContainer(
+      parent: root,
+      overrides: [overriden],
+    );
+
+    final instance1 = provider(0);
+    var count = container.read(instance1);
+    expect(count, 1);
+    await container.pump();
```
Comment (rrousselGit) [inline packages/riverpod/test/framework/auto_dispose_test.dart:552]:
```suggestion

    await container.pump();
```

File: packages/riverpod/test/framework/auto_dispose_test.dart [hunk 4]
```diff
@@ -529,6 +529,33 @@ final alwaysAlive = Provider((ref) {
...
+    });
+
+    final overriden = provider.overrideWithProvider(override);
+    final root = createContainer();
+    final container = createContainer(
+      parent: root,
+      overrides: [overriden],
+    );
+
+    final instance1 = provider(0);
```
Comment (rrousselGit) [inline packages/riverpod/test/framework/auto_dispose_test.dart:549]: Again, don't create variables for things like this.

File: packages/riverpod/test/framework/auto_dispose_test.dart [hunk 5]
```diff
@@ -529,6 +529,33 @@ final alwaysAlive = Provider((ref) {
...
+    var count = container.read(instance1);
+    expect(count, 1);
+    await container.pump();
+
+    final instance2 = provider(0);
```
Comment (rrousselGit) [inline packages/riverpod/test/framework/auto_dispose_test.dart:554]: Same here

File: packages/riverpod/test/framework/auto_dispose_test.dart [hunk 6]
```diff
@@ -529,6 +529,35 @@ final alwaysAlive = Provider((ref) {
...
   });
 
+  test(
+      'supports disposing of overridden families',
+      () async {
+    // Regression test for https://github.com/rrousselGit/riverpod/issues/2480
+    final provider = Provider.autoDispose.family<int, int>((ref, _) => -1);
+
+    var constructionCount = 0;
+    final override = Provider.autoDispose.family<int, int>((ref, _) {
```
Comment (rrousselGit) [inline packages/riverpod/test/framework/auto_dispose_test.dart:539]: Same here, no variable for such a thing

File: packages/riverpod/test/framework/auto_dispose_test.dart (hunk 2 above)
Comment (jeiea) [reply packages/riverpod/test/framework/auto_dispose_test.dart]: I forgot to remove them after inspecting. I'll close conversations that you've fixed all.

**Pull request merged**